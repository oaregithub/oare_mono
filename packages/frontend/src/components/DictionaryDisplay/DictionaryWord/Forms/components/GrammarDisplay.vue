<template>
  <div>
    <v-tooltip bottom open-delay="800">
      <template #activator="{ on, attrs }">
        <v-btn
          v-if="allowEditing && canEditParseInfo"
          icon
          class="test-property-pencil edit-button mr-1"
          @click="editPropertiesDialog = true"
          small
          v-bind="attrs"
          v-on="on"
        >
          <v-icon size="20">mdi-pencil</v-icon>
        </v-btn>
      </template>
      <span>Edit Parse Info</span>
    </v-tooltip>
    <span class="mr-1" v-if="formGrammar === ''">(No parse info yet)</span>
    <span class="mr-1" v-else-if="formGrammar !== ''">({{ formGrammar }})</span>
    <oare-dialog
      v-if="allowEditing && canEditParseInfo"
      v-model="editPropertiesDialog"
      :title="`Edit Form Parse Information - ${form.form}`"
      :width="1000"
      :submitDisabled="!formComplete"
      submitText="Submit"
      closeOnSubmit
      @submit="updateFormProperties"
    >
      <add-properties
        :startingUuid="partOfSpeechValueUuid"
        requiredNodeValueName="Parse"
        @export-properties="setProperties($event)"
        @form-complete="formComplete = $event"
        :existingProperties="form.properties"
        :key="addPropertiesKey"
      />
    </oare-dialog>
  </div>
</template>

<script lang="ts">
import {
  defineComponent,
  PropType,
  computed,
  ref,
  watch,
  inject,
} from '@vue/composition-api';
import { DictionaryForm, ParseTreeProperty, Word } from '@oare/types';
import utils from '@/utils';
import AddProperties from '@/components/Properties/AddProperties.vue';
import { ReloadKey } from '../../index.vue';
import sl from '@/serviceLocator';

export default defineComponent({
  props: {
    word: {
      type: Object as PropType<Word>,
      required: false,
    },
    form: {
      type: Object as PropType<DictionaryForm>,
      required: true,
    },
    allowEditing: {
      type: Boolean,
      default: true,
    },
  },
  components: {
    AddProperties,
  },
  setup(props) {
    const server = sl.get('serverProxy');
    const actions = sl.get('globalActions');
    const store = sl.get('store');
    const reload = inject(ReloadKey);

    const editPropertiesDialog = ref(false);
    const formComplete = ref(false);

    const formGrammar = computed(() => utils.formGrammarString(props.form));

    const properties = ref<ParseTreeProperty[]>([]);
    const setProperties = (propertyList: ParseTreeProperty[]) => {
      properties.value = propertyList;
    };

    const partOfSpeechValueUuid = computed(() => {
      if (props.word) {
        const posProperties = props.word.properties.filter(
          prop => prop.variableName === 'Part of Speech'
        );
        return posProperties.length > 0
          ? posProperties[0].valueUuid
          : undefined;
      }
    });

    const addPropertiesKey = ref(false);
    watch(editPropertiesDialog, () => {
      if (editPropertiesDialog.value) {
        addPropertiesKey.value = !addPropertiesKey.value;
        properties.value = [];
      }
    });

    const updateFormProperties = async () => {
      try {
        await server.editPropertiesByReferenceUuid(
          props.form.uuid,
          properties.value
        );
        actions.showSnackbar(
          `Successfully updated form parse info for ${props.form.form}`
        );
        reload && reload();
      } catch (err) {
        actions.showErrorSnackbar(
          'Error updating form parse information.',
          err as Error
        );
      } finally {
        properties.value = [];
      }
    };

    const canEditParseInfo = computed(() =>
      store.hasPermission('EDIT_ITEM_PROPERTIES')
    );

    return {
      formGrammar,
      editPropertiesDialog,
      formComplete,
      properties,
      setProperties,
      partOfSpeechValueUuid,
      addPropertiesKey,
      updateFormProperties,
      canEditParseInfo,
    };
  },
});
</script>

<style scoped>
.edit-button {
  margin-top: -11px;
  margin-bottom: -7px;
}
</style>
