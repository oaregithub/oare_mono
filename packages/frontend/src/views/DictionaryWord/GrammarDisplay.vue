<template>
  <span class="mx-1" v-if="formGrammar !== ''">({{ formGrammar }})</span>
</template>

<script lang="ts">
import {
  defineComponent,
  PropType,
  computed,
  ref,
  Ref,
  reactive,
} from '@vue/composition-api';
import { DictionaryForm, SpellingText } from '@oare/types';
import { DataTableHeader } from 'vuetify';

export default defineComponent({
  props: {
    form: {
      type: Object as PropType<DictionaryForm>,
      required: true,
    },
  },
  setup({ form }) {
    const search = ref('');

    const formGrammar = computed(() => {
      let suffix = '';

      if (form.suffix) {
        suffix =
          form.suffix.persons.join('') +
          form.suffix.genders.join('') +
          form.suffix.grammaticalNumbers.join('') +
          form.suffix.cases.join('');
        if (form.clitics.includes('-ma')) {
          suffix += '-ma';
        }
      }
      return (
        form.stems.join('') +
        form.morphologicalForms
          .filter(mf => mf === 'stv.' || mf === 'inf')
          .join('') +
        form.tenses.join('') +
        ' ' +
        form.persons.join('') +
        form.genders.join('') +
        form.grammaticalNumbers.join('') +
        form.cases.join('') +
        ' ' +
        form.states.join('') +
        form.moods.join('') +
        form.clitics
          .map(clitic => {
            if (clitic === 'suf.') return '+';
            if (clitic === 'vent') return '+vent';
            return clitic;
          })
          .filter(clitic => clitic !== '-ma')
          .join('') +
        suffix
      ).trim();
    });

    return {
      formGrammar,
      search,
    };
  },
});
</script>
