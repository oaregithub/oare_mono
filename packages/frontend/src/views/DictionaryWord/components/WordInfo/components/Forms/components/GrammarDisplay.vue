<template>
  <div>
    <v-tooltip bottom open-delay="800">
      <template #activator="{ on, attrs }">
        <v-btn
          v-if="allowEditing && canEditParseInfo"
          icon
          class="test-property-pencil edit-button"
          @click="editPropertiesDialog = true"
          small
          v-bind="attrs"
          v-on="on"
        >
          <v-icon size="20">mdi-pencil</v-icon>
        </v-btn>
      </template>
      <span>Edit Parse Info</span>
    </v-tooltip>
    <span class="mr-1" v-if="formGrammar === '' && canEditParseInfo"
      >(No parse info yet)</span
    >
    <span class="mr-1" v-else-if="formGrammar !== ''">({{ formGrammar }})</span>
    <oare-dialog
      v-if="allowEditing && canEditParseInfo"
      v-model="editPropertiesDialog"
      :title="`Edit Form Parse Information - ${form.form}`"
      :width="1400"
      :submitDisabled="!formComplete"
      submitText="Submit"
      closeOnSubmit
      @submit="updateFormProperties"
    >
      <properties-tree
        :readonly="false"
        startingValueHierarchyUuid="b745f8d1-55f2-11eb-bf9e-024de1c1cc1d"
        @set-properties="setProperties($event)"
        @set-complete="formComplete = $event"
        :existingProperties="form.properties"
        :key="addPropertiesKey"
      />
    </oare-dialog>
  </div>
</template>

<script lang="ts">
import {
  defineComponent,
  PropType,
  computed,
  ref,
  watch,
  inject,
} from '@vue/composition-api';
import { AppliedProperty, DictionaryForm, Word } from '@oare/types';
import utils from '@/utils';
import { ReloadKey } from '../../../../../index.vue';
import sl from '@/serviceLocator';
import PropertiesTree from '@/components/Properties/PropertiesTree.vue';

export default defineComponent({
  props: {
    word: {
      type: Object as PropType<Word>,
      required: false,
    },
    form: {
      type: Object as PropType<DictionaryForm>,
      required: true,
    },
    allowEditing: {
      type: Boolean,
      default: true,
    },
  },
  components: {
    PropertiesTree,
  },
  setup(props) {
    const server = sl.get('serverProxy');
    const actions = sl.get('globalActions');
    const store = sl.get('store');
    const reload = inject(ReloadKey);

    const editPropertiesDialog = ref(false);
    const formComplete = ref(false);

    const formGrammar = computed(() => utils.formGrammarString(props.form));

    const properties = ref<AppliedProperty[]>([]);
    const setProperties = (propertyList: AppliedProperty[]) => {
      properties.value = propertyList;
    };

    const addPropertiesKey = ref(false);
    watch(editPropertiesDialog, () => {
      if (editPropertiesDialog.value) {
        addPropertiesKey.value = !addPropertiesKey.value;
        properties.value = [];
      }
    });

    const updateFormProperties = async () => {
      try {
        await server.editPropertiesByReferenceUuid(
          props.form.uuid,
          properties.value,
          props.word ? props.word.uuid : undefined
        );
        actions.showSnackbar(
          `Successfully updated form parse info for ${props.form.form}`
        );
        reload && reload();
      } catch (err) {
        actions.showErrorSnackbar(
          'Error updating form parse information.',
          err as Error
        );
      } finally {
        properties.value = [];
      }
    };

    const canEditParseInfo = computed(() =>
      store.hasPermission('EDIT_ITEM_PROPERTIES')
    );

    return {
      formGrammar,
      editPropertiesDialog,
      formComplete,
      properties,
      setProperties,
      addPropertiesKey,
      updateFormProperties,
      canEditParseInfo,
    };
  },
});
</script>

<style scoped>
.edit-button {
  margin-top: -11px;
  margin-bottom: -7px;
}
</style>
