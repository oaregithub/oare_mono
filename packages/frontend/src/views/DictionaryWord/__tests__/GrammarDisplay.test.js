import Vuetify from 'vuetify';
import VueCompositionApi from '@vue/composition-api';
import { mount, createLocalVue } from '@vue/test-utils';
import GrammarDisplay from '../GrammarDisplay.vue';

const vuetify = new Vuetify();
const localVue = createLocalVue();
localVue.use(VueCompositionApi);

describe('GrammarDisplay test', () => {
  const mockForm = {
    form: 'Test-form',
    stems: [],
    tenses: [],
    persons: [],
    genders: [],
    grammaticalNumbers: [],
    cases: [],
    states: [],
    moods: [],
    clitics: [],
    morphologicalForms: [],
    suffix: null,
    spellings: [
      {
        uuid: 'spelling-uuid',
        spelling: 'spelling',
        texts: [
          {
            uuid: 'text-uuid',
            text: 'text',
          },
        ],
      },
    ],
  };
  const createWrapper = () =>
    mount(GrammarDisplay, {
      vuetify,
      localVue,
      propsData: {
        form: mockForm,
      },
      stubs: ['router-link'],
    });

  it('matches snapshot', () => {
    expect(createWrapper()).toMatchSnapshot();
  });

  it('shows number of texts next to spelling', () => {
    const wrapper = createWrapper();
    expect(Number(wrapper.find('.test-num-texts').text())).toBe(
      mockForm.spellings[0].texts.length
    );
  });

  it('shows dialog when clicking on number of texts', async () => {
    const wrapper = createWrapper();
    await wrapper.find('.test-num-texts').trigger('click');
    expect(wrapper.find('.test-dialog-title').text()).toBe(
      `Texts for ${mockForm.spellings[0].spelling}`
    );
  });

  it('shows texts associated with spelling', async () => {
    const wrapper = createWrapper();
    await wrapper.find('.test-num-texts').trigger('click');
    expect(wrapper.find('.test-text').text()).toBe(
      mockForm.spellings[0].texts[0].text
    );
  });
});
