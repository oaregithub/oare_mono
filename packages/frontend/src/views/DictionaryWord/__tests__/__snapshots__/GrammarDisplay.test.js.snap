// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GrammarDisplay test matches snapshot 1`] = `
<div>
  <!----> <span><span>spelling</span> <span>
      (<a class="test-num-texts">1</a>)</span>
  <!----></span>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</div>
`;
